Mage.Verify/src/test/java/mage/verify/VerifyCardDataTest.java

```diff
@@ -1646,19 +1651,102 @@ private void checkCardCanBeCopied(Card card1) {
-                Effect effect1 = ability1.getEffects().get(j);
-                Effect effect2 = ability2.getEffects().get(j);
-                if (!effect1.getClass().equals(effect2.getClass())) {
-                    fail(card1, "copy", "miss copy in effect " + effect1.getClass().getName());
+            // Only recurse on those objects
+            if (obj1 instanceof MageObject || obj1 instanceof Filter || obj1 instanceof Condition || obj1 instanceof Effect
+                    || obj1 instanceof Ability || obj1 instanceof Mana || obj1 instanceof Cost || obj1 instanceof DynamicValue
+                    || obj1 instanceof Choice || obj1 instanceof TargetPointer) {
+                //System.out.println(msg);
+                Class class1 = obj1.getClass();
+                Class class2 = obj2.getClass();
+                do {
+                    if (class1 == null && class2 == null) {
+                        return;
+                    }
+                    if (class1 == null || class2 == null) {
+                        fail(originalCard, "copy", "not same class for " + msg + "<" + obj1.getClass() + ">" + "]");
+                        return;
+                    }
+                    List<Field> ability2Fields = Arrays.stream(class2.getDeclaredFields()).collect(Collectors.toList());
+                    int fieldIndex = 0;
+                    for (Field field1 : class1.getDeclaredFields()) {
+                        Field field2 = ability2Fields.get(fieldIndex);
+                        field1.setAccessible(true);
+                        field2.setAccessible(true);
+                        try {
+                            Object value1 = field1.get(obj1);
+                            Object value2 = field2.get(obj2);
+
+                            boolean doFieldRecurse = true;
+                            if (class1 == CardImpl.class) {
+                                if (field1.getName() == "spellAbility") {
+                                    compareClassRecursive(((CardImpl) obj1).getSpellAbility(), ((CardImpl) obj2).getSpellAbility(), originalCard, msg + "<" + obj1.getClass() + ">" + "::" + field1.getName(), maxDepth - 1);
+                                    doFieldRecurse = false;
+                                } else if (field1.getName() == "meldsToCard") {
+                                    //compareClassRecursive(((CardImpl) obj1).getMeldsToCard(), ((CardImpl) obj2).getMeldsToCard(), originalCard, msg + "::" + field1.getName(), maxDepth - 1);
+                                    doFieldRecurse = false;
+                                } else if (field1.getName() == "secondSideCard") {
+                                    //compareClassRecursive(((CardImpl) obj1).getSecondCardFace(), ((CardImpl) obj2).getSecondCardFace(), originalCard, msg + "::" + field1.getName(), maxDepth - 1);
+                                    doFieldRecurse = false;
+                                }
```

Comment:
If someone has an idea how to go into the second card/meld face, but with some security to not loop back into the initial card, I could use that here. For now those are just skipped.

Reply:
Possible solution:
- compareClass with checkedClasses param as set and and useRecursive param as boolean;
- part 1: check object class;
- part 2: check all fields by compareClass call:
  * if not useRecusive then return;
  * cycle all fields:
    * if checkedClasses contains a field's class then skip it;
    * if field's class inheried from Card then add it to checkedClasses;
    * call compareClass with useRecursive = true;